fix encoding & decoding

Stebalien opened the pull request:
1. Correctly encode & decode positive signed integers.
2. Correctly handle integer overflow everywhere.
3. Correctly handle all tag types (can be 0-4 bytes).
4. Always encode floats as f64 (by DagCbor spec).
5. Enforce minimal integer encoding.
6. Reject almost everything that's not DagCbor.

To still consider:

1. Be stricter when skipping & scanning for refs:
    1. Validate integers.
    2. Validate floats.
    3. Validate utf8 in strings.
    4. Reject unrecognized tags.
2. Maybe allow on decode:
    1. Undefined.
    2. f16 decoding.
    3. nan, infinity, etc.
3. Maybe forbid on decode:
    1. f16 & f32 on decode.

Stebalien committed:
fix encoding & decoding

1. Correctly encode & decode positive signed integers.
2. Correctly handle integer overflow everywhere.
3. Correctly handle all tag types (can be 0-4 bytes).
4. Enforce minimal integer encoding.
5. Reject almost everything that's not DagCbor.

To still consider:

1. Be stricter when skipping & scanning for refs:
    1. Validate integers.
    2. Validate floats.
    3. Validate utf8 in strings.
    4. Reject unrecognized tags.
2. Maybe allow on decode:
    1. Undefined.
    2. f16 decoding.
    3. nan, infinity, etc.
3. Maybe forbid on decode:
    1. f16 & f32 on decode.

dvc94ch reviewed (approved):
Lgtm, I don't have utility for this anymore and actyx doesn't care.

vmx reviewed (commented):
The changes look good. Once CI is happy it's good to go.

Stebalien committed:
export and cleanup major helpers

We need these exported for the derive-cbor logic, so we might as well
cleanup a bit.

Stebalien committed:
fix cbor derive decoding

Also fixes some bugs where we'd make assumptions about having fewer than
24 fields.

Stebalien committed:
fix lints

vmx reviewed (commented):
Things look good. The only thing I'd like to mention is that you are using glob/star imports for the `cbor` module. I prefer explicit imports as it makes the code easier to grep in. I also find it easier to reason about it while reading, when I know where things come from.

So I suggest changing the `use cbor:*` imports to explicitly listing what gets imported.
